<!DOCTYPE html>
<html data-color-mode="light" data-dark-theme="dark" data-light-theme="light" lang="zh-CN">
<head>
    <meta content="text/html; charset=utf-8" http-equiv="content-type" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <link href='https://mirrors.sustech.edu.cn/cdnjs/ajax/libs/Primer/21.0.7/primer.css' rel='stylesheet' />
    
    <link rel="icon" href="https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/header.JPG"><script>
        let theme = localStorage.getItem("meek_theme") || "light";
        document.documentElement.setAttribute("data-color-mode", theme);
    </script>
<meta name="description" content="# 写在前面
1. 分析基于 [OrthoFinder](https://github.com/davidemms/OrthoFinder) 结果, 这个工具网上教程比较多了, 这里直接使用结果文件;
2. 部分代码参考[发表的葡萄泛基因组文章](https://github.com/NotebookOFXiaoMing/SuperPangenomeofGrapevines);
3. 本质是一个 R 脚本, 可视化部分使用 `ggplot2`, 方便大家自行修改;
# 接口说明
本工具包含以下接口:
- `-i`: OrthoFinder 结果中 `Orthogroups/Orthogroups.GeneCount.tsv` 文件;
- `-o`: 输出文件目录, 文件名已经在脚本中指定, 如果目录不存在**不会**创建;
- `-s`: 统计 pan/core 基因集时模拟不放回抽取的次数, 默认 100;
- `--core`: 核心基因集系数, 默认为 1;
- `--softcore`: 软核心基因集系数, 默认 0.9;
- `--private`: 私有基因集系数, 默认 0.01, 最小为 1 个样本;

特别说明一下后面三个参数, 脚本自动统计输入的样本数/基因集数, 在这个基础上, 乘以相应的系数用于基因集判断, 对于默认参数, 则:
- 出现在全部样本的基因家族视为 core;
- 出现在至少 90% 样本的基因家族视为 softcore;
- 仅有 1%样本(或仅有一个样本) 支持的基因家族视为 private;
- 其余基因家族视为 dispensable;
# 结果说明
主要结果包括三个图两个表:
![PanCoreGeneFamilies](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/PanCoreGeneFamilies.png)
![GeneFamiliesFrequency](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/GeneFamiliesFrequency.png)
![ClassficationDonut](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/ClassficationDonut.png)
主要统计表:
![ClassficationSummary](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/ClassficationSummary.png)

![GeneFamiliesClassfication](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main//_images/GeneFamiliesClassfication.png)

# 脚本使用
脚本已经发布在 [GitHub](https://github.com/WangZhSi/Bioinformatics_tool/tree/main/orthoPanCore_visualization) 上, 欢迎大家使用.。">
<meta property="og:title" content="泛基因集的识别与可视化">
<meta property="og:description" content="# 写在前面
1. 分析基于 [OrthoFinder](https://github.com/davidemms/OrthoFinder) 结果, 这个工具网上教程比较多了, 这里直接使用结果文件;
2. 部分代码参考[发表的葡萄泛基因组文章](https://github.com/NotebookOFXiaoMing/SuperPangenomeofGrapevines);
3. 本质是一个 R 脚本, 可视化部分使用 `ggplot2`, 方便大家自行修改;
# 接口说明
本工具包含以下接口:
- `-i`: OrthoFinder 结果中 `Orthogroups/Orthogroups.GeneCount.tsv` 文件;
- `-o`: 输出文件目录, 文件名已经在脚本中指定, 如果目录不存在**不会**创建;
- `-s`: 统计 pan/core 基因集时模拟不放回抽取的次数, 默认 100;
- `--core`: 核心基因集系数, 默认为 1;
- `--softcore`: 软核心基因集系数, 默认 0.9;
- `--private`: 私有基因集系数, 默认 0.01, 最小为 1 个样本;

特别说明一下后面三个参数, 脚本自动统计输入的样本数/基因集数, 在这个基础上, 乘以相应的系数用于基因集判断, 对于默认参数, 则:
- 出现在全部样本的基因家族视为 core;
- 出现在至少 90% 样本的基因家族视为 softcore;
- 仅有 1%样本(或仅有一个样本) 支持的基因家族视为 private;
- 其余基因家族视为 dispensable;
# 结果说明
主要结果包括三个图两个表:
![PanCoreGeneFamilies](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/PanCoreGeneFamilies.png)
![GeneFamiliesFrequency](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/GeneFamiliesFrequency.png)
![ClassficationDonut](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/ClassficationDonut.png)
主要统计表:
![ClassficationSummary](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/ClassficationSummary.png)

![GeneFamiliesClassfication](https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main//_images/GeneFamiliesClassfication.png)

# 脚本使用
脚本已经发布在 [GitHub](https://github.com/WangZhSi/Bioinformatics_tool/tree/main/orthoPanCore_visualization) 上, 欢迎大家使用.。">
<meta property="og:type" content="article">
<meta property="og:url" content="https://WangZhSi.github.io/post/fan-ji-yin-ji-de-shi-bie-yu-ke-shi-hua.html">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/header.JPG">
<title>泛基因集的识别与可视化</title>



</head>
<style>
body{box-sizing: border-box;min-width: 200px;max-width: 900px;margin: 20px auto;padding: 45px;font-size: 16px;font-family: sans-serif;line-height: 1.25;}
#header{display:flex;padding-bottom:8px;border-bottom: 1px solid var(--borderColor-muted, var(--color-border-muted));margin-bottom: 16px;}
#footer {margin-top:64px; text-align: center;font-size: small;}

</style>

<style>
.postTitle{margin: auto 0;font-size:40px;font-weight:bold;}
.title-right{display:flex;margin:auto 0 0 auto;}
.title-right .circle{padding: 14px 16px;margin-right:8px;}
#postBody{border-bottom: 1px solid var(--color-border-default);padding-bottom:36px;}
#postBody hr{height:2px;}
#cmButton{height:48px;margin-top:48px;}
#comments{margin-top:64px;}
.g-emoji{font-size:24px;}
@media (max-width: 600px) {
    body {padding: 8px;}
    .postTitle{font-size:24px;}
}

</style>
<link href='https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_assets/style.css' rel='stylesheet' /><style>.markdown-body{font-family: myFira, myLXGW, Arial;}</style><style>#postBody img {display: block; margin-left: auto; margin-right: auto;}</style>



<body>
    <div id="header">
<h1 class="postTitle">泛基因集的识别与可视化</h1>
<div class="title-right">
    <a href="https://WangZhSi.github.io" id="buttonHome" class="btn btn-invisible circle" title="首页">
        <svg class="octicon" width="16" height="16">
            <path id="pathHome" fill-rule="evenodd"></path>
        </svg>
    </a>
    
    <a href="https://github.com/WangZhSi/WangZhSi.github.io/issues/11" target="_blank" class="btn btn-invisible circle" title="Issue">
        <svg class="octicon" width="16" height="16">
            <path id="pathIssue" fill-rule="evenodd"></path>
        </svg>
    </a>
    

    <a class="btn btn-invisible circle" onclick="modeSwitch();" title="切换主题">
        <svg class="octicon" width="16" height="16" >
            <path id="themeSwitch" fill-rule="evenodd"></path>
        </svg>
    </a>

</div>
</div>
    <div id="content">
<div class="markdown-body" id="postBody"><h1>写在前面</h1>
<ol>
<li>分析基于 <a href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a> 结果, 这个工具网上教程比较多了, 这里直接使用结果文件;</li>
<li>部分代码参考<a href="https://github.com/NotebookOFXiaoMing/SuperPangenomeofGrapevines">发表的葡萄泛基因组文章</a>;</li>
<li>本质是一个 R 脚本, 可视化部分使用 <code class="notranslate">ggplot2</code>, 方便大家自行修改;</li>
</ol>
<h1>接口说明</h1>
<p>本工具包含以下接口:</p>
<ul>
<li><code class="notranslate">-i</code>: OrthoFinder 结果中 <code class="notranslate">Orthogroups/Orthogroups.GeneCount.tsv</code> 文件;</li>
<li><code class="notranslate">-o</code>: 输出文件目录, 文件名已经在脚本中指定, 如果目录不存在<strong>不会</strong>创建;</li>
<li><code class="notranslate">-s</code>: 统计 pan/core 基因集时模拟不放回抽取的次数, 默认 100;</li>
<li><code class="notranslate">--core</code>: 核心基因集系数, 默认为 1;</li>
<li><code class="notranslate">--softcore</code>: 软核心基因集系数, 默认 0.9;</li>
<li><code class="notranslate">--private</code>: 私有基因集系数, 默认 0.01, 最小为 1 个样本;</li>
</ul>
<p>特别说明一下后面三个参数, 脚本自动统计输入的样本数/基因集数, 在这个基础上, 乘以相应的系数用于基因集判断, 对于默认参数, 则:</p>
<ul>
<li>出现在全部样本的基因家族视为 core;</li>
<li>出现在至少 90% 样本的基因家族视为 softcore;</li>
<li>仅有 1%样本(或仅有一个样本) 支持的基因家族视为 private;</li>
<li>其余基因家族视为 dispensable;</li>
</ul>
<h1>结果说明</h1>
<p>主要结果包括三个图两个表:<br>
<a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/b30fc079c27366e9b4a55fa5bfd3998db765e776fbd8930648ceb14c14797884/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f50616e436f726547656e6546616d696c6965732e706e67"><img src="https://camo.githubusercontent.com/b30fc079c27366e9b4a55fa5bfd3998db765e776fbd8930648ceb14c14797884/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f50616e436f726547656e6546616d696c6965732e706e67" alt="PanCoreGeneFamilies" data-canonical-src="https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/PanCoreGeneFamilies.png" style="max-width: 100%; height: auto;"></a><br>
<a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/9318a8216138842d90346a07692953a4acabaad952e61f2e4f7a0888c8c7077e/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f47656e6546616d696c6965734672657175656e63792e706e67"><img src="https://camo.githubusercontent.com/9318a8216138842d90346a07692953a4acabaad952e61f2e4f7a0888c8c7077e/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f47656e6546616d696c6965734672657175656e63792e706e67" alt="GeneFamiliesFrequency" data-canonical-src="https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/GeneFamiliesFrequency.png" style="max-width: 100%; height: auto;"></a><br>
<a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/417538922ef7003fed8f964d0193caf0f32acb1b954e8ff537e01b158a9d1e43/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f436c6173736669636174696f6e446f6e75742e706e67"><img src="https://camo.githubusercontent.com/417538922ef7003fed8f964d0193caf0f32acb1b954e8ff537e01b158a9d1e43/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f436c6173736669636174696f6e446f6e75742e706e67" alt="ClassficationDonut" data-canonical-src="https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/ClassficationDonut.png" style="max-width: 100%; height: auto;"></a><br>
主要统计表:<br>
<a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/42d876b18080ba2a88061536ad84c3542646f5bfabc3c3b9e25f5ce5c6c876ef/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f436c6173736669636174696f6e53756d6d6172792e706e67"><img src="https://camo.githubusercontent.com/42d876b18080ba2a88061536ad84c3542646f5bfabc3c3b9e25f5ce5c6c876ef/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f5f696d616765732f436c6173736669636174696f6e53756d6d6172792e706e67" alt="ClassficationSummary" data-canonical-src="https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main/_images/ClassficationSummary.png" style="max-width: 100%; height: auto;"></a></p>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/da563a5fb17f0de675c7d69d5af21503fe66ee178d505c43c33b0ac20be70dba/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f2f5f696d616765732f47656e6546616d696c696573436c6173736669636174696f6e2e706e67"><img src="https://camo.githubusercontent.com/da563a5fb17f0de675c7d69d5af21503fe66ee178d505c43c33b0ac20be70dba/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f57616e675a6853692f57616e675a6853692e6769746875622e696f406d61696e2f2f5f696d616765732f47656e6546616d696c696573436c6173736669636174696f6e2e706e67" alt="GeneFamiliesClassfication" data-canonical-src="https://cdn.jsdelivr.net/gh/WangZhSi/WangZhSi.github.io@main//_images/GeneFamiliesClassfication.png" style="max-width: 100%; height: auto;"></a></p>
<h1>脚本使用</h1>
<p>脚本已经发布在 <a href="https://github.com/WangZhSi/Bioinformatics_tool/tree/main/orthoPanCore_visualization">GitHub</a> 上, 欢迎大家使用.</p></div>
<div style="font-size:small;margin-top:8px;float:right;"></div>

<button class="btn btn-block" type="button" onclick="openComments()" id="cmButton">评论</button>
<div class="comments" id="comments"></div>

</div>
    <div id="footer"><div id="footer1">Copyright © <span id="copyrightYear"></span> <a href="https://WangZhSi.github.io">勺水清欢</a></div>
<div id="footer2">
    <span id="runday"></span><span>Powered by <a href="https://meekdai.com/Gmeek.html" target="_blank">Gmeek</a></span>
</div>

<script>
var now=new Date();
document.getElementById("copyrightYear").innerHTML=now.getFullYear();

if("05/23/2024"!=""){
    var startSite=new Date("05/23/2024");
    var diff=now.getTime()-startSite.getTime();
    var diffDay=Math.floor(diff/(1000*60*60*24));
    document.getElementById("runday").innerHTML="网站运行"+diffDay+"天"+" • ";
}
</script></div>
</body>
<script>
var IconList={'sun': 'M8 10.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5zM8 12a4 4 0 100-8 4 4 0 000 8zM8 0a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0V.75A.75.75 0 018 0zm0 13a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0v-1.5A.75.75 0 018 13zM2.343 2.343a.75.75 0 011.061 0l1.06 1.061a.75.75 0 01-1.06 1.06l-1.06-1.06a.75.75 0 010-1.06zm9.193 9.193a.75.75 0 011.06 0l1.061 1.06a.75.75 0 01-1.06 1.061l-1.061-1.06a.75.75 0 010-1.061zM16 8a.75.75 0 01-.75.75h-1.5a.75.75 0 010-1.5h1.5A.75.75 0 0116 8zM3 8a.75.75 0 01-.75.75H.75a.75.75 0 010-1.5h1.5A.75.75 0 013 8zm10.657-5.657a.75.75 0 010 1.061l-1.061 1.06a.75.75 0 11-1.06-1.06l1.06-1.06a.75.75 0 011.06 0zm-9.193 9.193a.75.75 0 010 1.06l-1.06 1.061a.75.75 0 11-1.061-1.06l1.06-1.061a.75.75 0 011.061 0z', 'moon': 'M9.598 1.591a.75.75 0 01.785-.175 7 7 0 11-8.967 8.967.75.75 0 01.961-.96 5.5 5.5 0 007.046-7.046.75.75 0 01.175-.786zm1.616 1.945a7 7 0 01-7.678 7.678 5.5 5.5 0 107.678-7.678z', 'sync': 'M1.705 8.005a.75.75 0 0 1 .834.656 5.5 5.5 0 0 0 9.592 2.97l-1.204-1.204a.25.25 0 0 1 .177-.427h3.646a.25.25 0 0 1 .25.25v3.646a.25.25 0 0 1-.427.177l-1.38-1.38A7.002 7.002 0 0 1 1.05 8.84a.75.75 0 0 1 .656-.834ZM8 2.5a5.487 5.487 0 0 0-4.131 1.869l1.204 1.204A.25.25 0 0 1 4.896 6H1.25A.25.25 0 0 1 1 5.75V2.104a.25.25 0 0 1 .427-.177l1.38 1.38A7.002 7.002 0 0 1 14.95 7.16a.75.75 0 0 1-1.49.178A5.5 5.5 0 0 0 8 2.5Z', 'home': 'M6.906.664a1.749 1.749 0 0 1 2.187 0l5.25 4.2c.415.332.657.835.657 1.367v7.019A1.75 1.75 0 0 1 13.25 15h-3.5a.75.75 0 0 1-.75-.75V9H7v5.25a.75.75 0 0 1-.75.75h-3.5A1.75 1.75 0 0 1 1 13.25V6.23c0-.531.242-1.034.657-1.366l5.25-4.2Zm1.25 1.171a.25.25 0 0 0-.312 0l-5.25 4.2a.25.25 0 0 0-.094.196v7.019c0 .138.112.25.25.25H5.5V8.25a.75.75 0 0 1 .75-.75h3.5a.75.75 0 0 1 .75.75v5.25h2.75a.25.25 0 0 0 .25-.25V6.23a.25.25 0 0 0-.094-.195Z', 'github': 'M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z'};
var utterancesLoad=0;

let themeSettings={
    "dark": ["dark","moon","#00f0ff","dark-blue"],
    "light": ["light","sun","#ff5000","github-light"],
    "auto": ["auto","sync","","preferred-color-scheme"]
};
function changeTheme(mode, icon, color, utheme){
    document.documentElement.setAttribute("data-color-mode",mode);
    document.getElementById("themeSwitch").setAttribute("d",value=IconList[icon]);
    document.getElementById("themeSwitch").parentNode.style.color=color;
    if(utterancesLoad==1){utterancesTheme(utheme);}
}
function modeSwitch(){
    let currentMode=document.documentElement.getAttribute('data-color-mode');
    let newMode = currentMode === "light" ? "dark" : currentMode === "dark" ? "auto" : "light";
    localStorage.setItem("meek_theme", newMode);
    if(themeSettings[newMode]){
        changeTheme(...themeSettings[newMode]);
    }
}
function utterancesTheme(theme){
    const message={type:'set-theme',theme: theme};
    const iframe=document.getElementsByClassName('utterances-frame')[0];
    iframe.contentWindow.postMessage(message,'https://utteranc.es');
}
if(themeSettings[theme]){changeTheme(...themeSettings[theme]);}
console.log("\n %c Gmeek last https://github.com/Meekdai/Gmeek \n","padding:5px 0;background:#02d81d;color:#fff");
</script>

<script>
document.getElementById("pathHome").setAttribute("d",IconList["home"]);
document.getElementById("pathIssue").setAttribute("d",IconList["github"]);



function openComments(){
    cm=document.getElementById("comments");
    cmButton=document.getElementById("cmButton");
    cmButton.innerHTML="loading";
    span=document.createElement("span");
    span.setAttribute("class","AnimatedEllipsis");
    cmButton.appendChild(span);

    script=document.createElement("script");
    script.setAttribute("src","https://utteranc.es/client.js");
    script.setAttribute("repo","WangZhSi/WangZhSi.github.io");
    script.setAttribute("issue-term","title");
    
    if(localStorage.getItem("meek_theme")=="dark"){script.setAttribute("theme","dark-blue");}
    else if(localStorage.getItem("meek_theme")=="light") {script.setAttribute("theme","github-light");}
    else{script.setAttribute("theme","preferred-color-scheme");}
    
    script.setAttribute("crossorigin","anonymous");
    script.setAttribute("async","");
    cm.appendChild(script);

    int=self.setInterval("iFrameLoading()",200);
}

function iFrameLoading(){
    var utterances=document.getElementsByClassName('utterances');
    if(utterances.length==1){
        if(utterances[0].style.height!=""){
            utterancesLoad=1;
            int=window.clearInterval(int);
            document.getElementById("cmButton").style.display="none";
            console.log("utterances Load OK");
        }
    }
}



</script>


</html>
